### PR TITLE
Publish replication task to Kafka after reading from replicator queue

### PR DESCRIPTION
**Pull request by samarabbas**
Created QueueProcessorBase which has common logic used by both transfer
queue procesor and replication queue processor.

Created replicationQueueProcessor which processes replication task from
replicator queue and publishes it to Kafka.

Bootstrap logic for replicationQueueProcessor to host it within
historyEngine.  Also bootstrap logic to pass in kafka publisher to
history engine.

History engine changes to update replication state on
StartWorkflowExecution and UpdateWorkflowExecution by updating the
failover version from domain and writing it to mutable state.

### CODE REVIEW

File: service/history/mutableStateBuilder.go [hunk 1]
```diff
@@ -230,6 +230,12 @@ func (e *mutableStateBuilder) FlushBufferedEvents() error {
 	return nil
 }
 
+func (e *mutableStateBuilder) ApplyReplicationStateUpdates(failoverVersion int64) {
+	e.replicationState.CurrentVersion = failoverVersion
+	e.replicationState.LastWriteVersion = failoverVersion
+	e.replicationState.LastWriteEventID = e.hBuilder.nextEventID - 1
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go]: can we make this attr NextEventID? so no need to do the -1 operation

File: service/history/queueProcessor.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+			p.processWithRetry(task)
+		}
+	}
+}
+
+func (p *queueProcessorBase) processWithRetry(task queueTaskInfo) {
+	p.logger.Debugf("Processing task: %v, type: %v", task.GetTaskID(), task.GetTaskType())
+ProcessRetryLoop:
+	for retryCount := 1; retryCount <= 100; retryCount++ {
```
Comment (wxing1292) [inline service/history/queueProcessor.go]: retryCount should be configurable, ie. a const / var or input parameter

File: service/history/queueProcessor.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+func (a *ackManager) updateAckLevel() {
+	a.metricsClient.IncCounter(a.options.MetricScope, metrics.AckLevelUpdateCounter)
+	initialAckLevel := a.ackLevel
+	updatedAckLevel := a.ackLevel
+	a.Lock()
+MoveAckLevelLoop:
+	for current := a.ackLevel + 1; current <= a.readLevel; current++ {
+		if acked, ok := a.outstandingTasks[current]; ok {
```
Comment (wxing1292) [inline service/history/queueProcessor.go]: we should probably panic if !ok, i.e. key not found.

File: service/history/queueProcessor.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+		isStopped  int32
+		shutdownWG sync.WaitGroup
+		shutdownCh chan struct{}
+	}
+
+	// ackManager is created by QueueProcessor to keep track of the queue ackLevel for the shard.
+	// It keeps track of read level when dispatching tasks to processor and maintains a map of outstanding tasks.
+	// Outstanding tasks map uses the task id sequencer as the key, which is used by updateAckLevel to move the ack level
+	// for the shard when all preceding tasks are acknowledged.
+	ackManager struct {
```
Comment (wxing1292) [inline service/history/queueProcessor.go]: can we separate this ack manager into a dedicated file? that will be easier to add unit test and have separation of logic which will make the code more readable

File: service/history/mutableStateBuilder.go (hunk 1 above)
Comment (samarabbas) [reply service/history/mutableStateBuilder.go]: I'll put a TODO comment for this now.  Replication protocol currently uses LastWriteEventID so I don't want to mess with the terminology.  But I see your point.  I'll rename this when I start work on the replication protocol part of changes.

File: service/history/queueProcessor.go (hunk 4 above)
Comment (samarabbas) [reply service/history/queueProcessor.go]: For now it is not used by any other component and only implementation detail of queue processor.  Let me put a comment to move this outside.

File: service/history/queueProcessor.go (hunk 2 above)
Comment (samarabbas) [reply service/history/queueProcessor.go]: Good point.  Let me make this change.

File: service/history/queueProcessor.go (hunk 3 above)
Comment (samarabbas) [reply service/history/queueProcessor.go]: Well for now I ported over the existing logic.  Let me add a comment address what happens if !ok.

**Pull request merged**